internal/sgx/sgx.go

```diff
@@ -608,14 +611,46 @@ func (ctx *SgxContext) generateQuote(pubKey pkcs11.ObjectHandle) ([]byte, error)
 
 	quotePubKey, err := ctx.p11Ctx.WrapKey(ctx.p11Session, []*pkcs11.Mechanism{m}, pkcs11.ObjectHandle(0), pubKey)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	offset := uint64(C.quote_offset(*(*C.CK_BYTE_PTR)(unsafe.Pointer(&quotePubKey))))
 	if offset <= 0 || offset >= uint64(len(quotePubKey)) {
-		return nil, fmt.Errorf("quote generation failure: invalid quote")
+		return nil, nil, fmt.Errorf("quote generation failure: invalid quote")
+	}
+
+	publicKey, err := parseQuotePublickey(quotePubKey[:offset])
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to parse quote public key: %v", err)
+	}
+
+	return quotePubKey[offset:], publicKey, nil
+}
+
+// parseQuotePublickey reconstruct the rsa public key
+// from received bytes, received bytes structure like this:
+// pubkey_params   |    ulExponentLen   |    ulModulusLen
+// need to slice ulExponentLen and ulModulusLen to
+// reconstruct pubkey according to the size of each item
+func parseQuotePublickey(pubkey []byte) (*rsa.PublicKey, error) {
+	paramsSize := uint64(C.params_size(*(*C.CK_BYTE_PTR)(unsafe.Pointer(&pubkey))))
+	exponentLen := uint64(C.ulExponentLen_offset(*(*C.CK_BYTE_PTR)(unsafe.Pointer(&pubkey))))
+
+	var bigExponent = new(big.Int)
+	bigExponent.SetBytes(pubkey[paramsSize : paramsSize+exponentLen])
+	if bigExponent.BitLen() > 32 || bigExponent.Sign() < 1 {
+		return nil, fmt.Errorf("malformed quote public key")
+	}
+	if bigExponent.Uint64() > uint64(math.MaxInt) {
+		return nil, fmt.Errorf("malformed quote public key: possible data loss in exponent value")
 	}
-	return quotePubKey[offset:], nil
+	exponent := int(bigExponent.Uint64())
+	var modulus = new(big.Int)
+	modulus.SetBytes(pubkey[paramsSize+exponentLen:])
```

Comment:
Is there still a need to check the length here? To prevent out-of-bounds error.

Reply:
Yes, the check is needed. I added it now.